xcursormgr dont apply scale on gsettings

gulafaran opened the pull request:
gtk scales the cursor size itself since its CSD so if we scale the size its gonna get double scaled. incorporate the scale into xcursormanager to keep track of it.

also xwayland expects a scale of 1 for its cursor aswell.

see #7283

MahouShoujoMivutilde commented:
Tried it, here are some things I've noticed:

* `cursor:allow_dumb_copy = yes` again makes it bigger than it should be in wayland (everywhere) and proportionally smaller in xwayland (everywhere)
* when `XCURSOR_SIZE` isn't set, `xwayland:force_zero_scaling = yes` isn't respected, and the cursor gets bigger than in wayland over **non gtk** xwayland apps force scaled at 1
* when `XCURSOR_SIZE` is set, the cursor stays the same size everywhere regardless of force_zero_scaling, which is how it should be

Xwayland fractional scaling was always shit, if that's what you mean. I'm not sure if people even use it considering how blurry it is.

Xwayland scale 2 is still blurry for me (maybe need to set something else, aside from `monitor=...,2`), cursor has some haze over it, but at least its scale is the same as wayland.

This is xwayland evince at scale 2

![xwayland-evince-scale-2](https://github.com/user-attachments/assets/3850b31b-69ab-4cb1-b88d-aaf20e447a40)


Here is how it looks on wayland

![native-evince-scale-2](https://github.com/user-attachments/assets/baf71b38-cb39-4b43-bbe9-3c31e64a8496)

gulafaran commented:
> Tried it, here are some things I've noticed:
> 
>     * `cursor:allow_dumb_copy = yes` again makes it bigger than it should be in wayland (everywhere) and proportionally smaller in xwayland (everywhere)
> 
>     * when `XCURSOR_SIZE` isn't set, `xwayland:force_zero_scaling = yes` isn't respected, and the cursor gets bigger than in wayland over **non gtk** xwayland apps force scaled at 1
> 
>     * when `XCURSOR_SIZE` is set, the cursor stays the same size everywhere regardless of force_zero_scaling, which is how it should be
> 
> 
> Xwayland fractional scaling was always shit, if that's what you mean. I'm not sure if people even use it considering how blurry it is.
> 
> Xwayland scale 2 is still blurry for me (maybe need to set something else, aside from `monitor=...,2`), cursor has some haze over it, but at least its scale is the same as wayland.
> 
> This is xwayland evince at scale 2
> 
> ![xwayland-evince-scale-2](https://private-user-images.githubusercontent.com/14999778/357567681-3850b31b-69ab-4cb1-b88d-aaf20e447a40.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1ODI0NjUsIm5iZiI6MTcyMzU4MjE2NSwicGF0aCI6Ii8xNDk5OTc3OC8zNTc1Njc2ODEtMzg1MGIzMWItNjlhYi00Y2IxLWI4OGQtYWFmMjBlNDQ3YTQwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDIwNDkyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0MWQ4NTM3NzllOWI1MWYwNWIzNzNjMjRmYzMyOTdhN2RkODlmY2ViMWNmYzUwZGQ1YmI3ZjFmNmY0YzUwZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.U0X848OUDfWGXYeir3B48F3SJR9FXyD8bdMixpWf9no)
> 
> Here is how it looks on wayland
> 
> ![native-evince-scale-2](https://private-user-images.githubusercontent.com/14999778/357567694-baf71b38-cb39-4b43-bbe9-3c31e64a8496.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1ODI0NjUsIm5iZiI6MTcyMzU4MjE2NSwicGF0aCI6Ii8xNDk5OTc3OC8zNTc1Njc2OTQtYmFmNzFiMzgtY2IzOS00YjQzLWJiZTktM2MzMWU2NGE4NDk2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDIwNDkyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlhNDkyN2VlNGVjNGRjYTVhY2IxNmQ1YTQzOTM2YzRkODY0MmEzZmU1OTRhNDE2YTNhNjc0MWI5NGEyMWRhYmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IQ8_L54gkSaUo3pobKJjcVdurB-VINmRWUF8vH-P-Jw)

thanks for the input, il see into what i can do. the issue im trying to figure out is this


https://github.com/user-attachments/assets/c5309704-d399-433c-ac8c-fb38ac9448b9

MahouShoujoMivutilde commented:
I don't experience this at any scale.

Are you sure it's not some Hyprland in Hyprland thing?

https://github.com/user-attachments/assets/85ef43a4-c806-4cb3-943c-9d80c3e1d3ab

gulafaran commented:
> I don't experience this at any scale.
> 
> Are you sure it's not some Hyprland in Hyprland thing?
> 
>  glxgears-scale-2-xwayland.mp4

Oh never considered that maybe

MahouShoujoMivutilde commented:
I had glxgears opened on other workspace, and it just crashed [hyprlandCrashReport115793.txt](https://github.com/user-attachments/files/16605418/hyprlandCrashReport115793.txt) (on my config).

I've changed scale without restarting glxgears, a bunch of times, tried to run Hyprland in Hyprland too (had a problem with keys being pressed twice, so closed it, and went to figure out that).

Tried to reproduce on minimal config 
[hdefault-min.conf.txt](https://github.com/user-attachments/files/16605605/hdefault-min.conf.txt) with some my keybindings, just opened glxgear and changed scale a bunch of times, and then went to write a shorter version of this comment, and it crashed. 
[hyprlandCrashReport9244.txt](https://github.com/user-attachments/files/16605618/hyprlandCrashReport9244.txt).

I'm not sure if it's even related to this MR to be honest, I've never tried to do that with glxgears.

gulafaran commented:
> I had glxgears opened on other workspace, and it just crashed [hyprlandCrashReport115793.txt](https://github.com/user-attachments/files/16605418/hyprlandCrashReport115793.txt) (on my config).
> 
> I've changed scale without restarting glxgears, a bunch of times, tried to run Hyprland in Hyprland too (had a problem with keys being pressed twice, so closed it, and went to figure out that).
> 
> Tried to reproduce on minimal config [hdefault-min.conf.txt](https://github.com/user-attachments/files/16605605/hdefault-min.conf.txt) with some my keybindings, just opened glxgear and changed scale a bunch of times, and then went to write a shorter version of this comment, and it crashed. [hyprlandCrashReport9244.txt](https://github.com/user-attachments/files/16605618/hyprlandCrashReport9244.txt).
> 
> I'm not sure if it's even related to this MR to be honest, I've never tried to do that with glxgears.

nah thats something else https://github.com/hyprwm/Hyprland/blob/3b4aabe04c7756fb0a70d78b6f0e701228f46345/src/events/Monitors.cpp#L101 one of these Debug::log lines caused the crash, most likely some null member access or a null pointer. why and how it happend im not sure but thats for another bughunt :p

MahouShoujoMivutilde commented:
Well, I'm too lazy do it, and it doesn't seem like it crashes if you don't mess with scale so much.

MahouShoujoMivutilde commented:
Hold on, do you have some variable refresh rate monitor?

I'm looking at glxgears log on your recording, and it's ~probably 240hz at the beginning, but then ~90 and ~120, and if the low fps in the middle can be dismissed as just it being not visible and rendering at minimal performance, the last one doesn't make sense - you're looking at it.

gulafaran commented:
> Hold on, do you have some variable refresh rate monitor?
> 
> I'm looking at glxgears log on your recording, and it's ~probably 240hz at the beginning, but then ~90 and ~120, and if the low fps in the middle can be dismissed as just it being not visible and rendering at minimal performance, the last one doesn't make sense - you're looking at it.
> 
> * I was trying to guess why it might have a cursor cut in half for you in xwayland.

im not sure nested fps matters with all them explicit sync/vrr/vsync turned on heh but yeah i have a variable refresh rate monitor

MahouShoujoMivutilde commented:
I'm not sure how it's rendered, and can it even render anything outside some cursor box, maybe it can. But image if it somehow constrained where it can render the cursor, and frames are always consistently somehow out of sync, but only for the cursor.

Try with variable refresh rate disabled, maybe it'll work.

MahouShoujoMivutilde commented:
Something is up with gtk4 apps

https://github.com/user-attachments/assets/e1f2d70c-8880-42be-b004-006a5f68431d

gulafaran commented:
> Something is up with gtk4 apps
> pavucontrol-gtk4.mp4
> 
> This is pavucontrol. Wayland.

that i mostly blame gtk4 and xcursors. client sided rendering bullcrap. all we are doing is setting a theme the shape it does itself

vaxerski commented:
request my review when ready

MahouShoujoMivutilde commented:
Yeah, it's gtk4.

gtk3 works fine with this MR, as does xwayland (aside from initial feedback about allow_dumb_copy / force_zero_scaling).

Even in sway, on scale 1 gtk4 renders cursor slightly bigger than it should.

https://github.com/user-attachments/assets/c4a4603f-5a0f-41c1-b3b3-1db2f9cdbdfc

And on scale 1.25 gtk3 is handled, xwayland is handled, but gtk4 is still bigger.

https://github.com/user-attachments/assets/cf9d6352-5cd3-459d-a7c5-45ec989f4bb8

(evince is gtk3 document viewer)

vaxerski reviewed (approved):
lgtm :)

b4shful commented:
@gulafaran nice work!!💖 

been following this issue and knowing how confusing the combination of cursors, Wayland and hidpi/display scaling seems to be, I thought "oh god, even when this is fixed I'll probably have to go through my dotfiles and make sure all my cursor settings are the right things"

Pleased to say that updating to apply your fix completely solved the issue without anything else needed on my side! As someone with a weird display and strange scaling, thank you I appreciate it a lot🫶

24bit192kHz commented:
Sorry, I'm really confused. I've tried doing all the things I saw here and here: https://github.com/hyprwm/Hyprland/issues/7283, but I'm only facing this issue in [sxiv]( https://github.com/xyb3rt/sxiv) and [nsxiv](https://github.com/nsxiv/nsxiv). The confusing part is that GIMP, Pavucontrol, and basically everything else on my system work fine.

https://github.com/user-attachments/assets/9a43d440-413f-40d3-960a-7a026e54d9f4

I reinstalled my system 4 days ago, so whatever changes you guys pushed, I should have them.

Also, I know it's off-topic, but why isn’t the Alacritty icon showing in Waybar?

MahouShoujoMivutilde commented:
This seems to be what I've [mentioned](https://github.com/hyprwm/Hyprland/pull/7316#issuecomment-2287111774) initially 

> when XCURSOR_SIZE isn't set, xwayland:force_zero_scaling = yes isn't respected, and the cursor gets bigger than in wayland over **non gtk** xwayland apps force scaled at 1

Nsxiv, which I use too, without issues, isn't gtk, while the rest you mentioned are gtk apps.

To "fix" add 

```ini
env = XCURSOR_SIZE,24
env = HYPRCURSOR_SIZE,24
```

and don't forget to restart hyprland for it to take effect.

izmyname commented:
> ```ini
> env = XCURSOR_SIZE,24
> ```

Didn't work for me, but the problem with giant adwaita cursor mostly happens in flatpak GTK apps. It wasn't the case somewhere a month ago.

A temporary solution for me is manually setting xcursor in nwg-look every time, my pc is booted.

izmyname commented:
UPD: setting XCURSOR_THEME and XCURSOR_SIZE in hyprland.conf seems to fixed my problem.
